### ISSUE TITLE
`app-js` sample throws exception

### ISSUE THREAD

**Issue by 05nelsonm** [arkivanov/Decompose#432]
Playing around with my own react app and I keep running into an uncaught exception after a page refresh attributed to the `UniqueId.kt` (a copy/paste of the `app-js` sample's). The same issue exists in the `Decompose/sample/app-js` project; wondering how to fix it :thinking: .

Steps to reproduce:
1. `./gradlew :sample:app-js:run`
2. Open browser console
3. Click around to different locations
4. Refresh the page
5. Cycle browser previous/forward buttons
6. Go back to step 2 and repeat until the exception is thrown

It crashes the app and then gets stuck until you refresh the page. I'm unsure if this has something to do with the `WebHistoryController` or not, so.

<details>
    <summary>Decompose/sample/app-js Logs</summary>

```
index.js:551 [webpack-dev-server] Live Reloading enabled.
Preconditions.kt:98 Uncaught IllegalStateException {message: 'Configurations must be unique', cause: undefined, name: 'IllegalStateException', stack: 'IllegalStateException: Configurations must be uniq…///./kotlin/Decompose-decompose-js-ir.js:3475:18)'}cause: undefinedmessage: "Configurations must be unique"name: "IllegalStateException"stack: "IllegalStateException: Configurations must be unique\n    at protoOf.switch_55pz6e_k$ (webpack-internal:///./kotlin/Decompose-decompose-js-ir.js:1548:13)\n    at protoOf.navigate_tfovv5_k$ (webpack-internal:///./kotlin/Decompose-decompose-js-ir.js:1525:10)\n    at eval (webpack-internal:///./kotlin/Decompose-decompose-js-ir.js:1281:18)\n    at drainLoop (webpack-internal:///./kotlin/Decompose-decompose-js-ir.js:394:9)\n    at protoOf.accept_kmf59w_k$ (webpack-internal:///./kotlin/Decompose-decompose-js-ir.js:424:5)\n    at protoOf.navigate_4i2940_k$ (webpack-internal:///./kotlin/Decompose-decompose-js-ir.js:2425:18)\n    at navigate (webpack-internal:///./kotlin/Decompose-decompose-js-ir.js:2469:19)\n    at protoOf.onPopState_a5hf3g_k$ (webpack-internal:///./kotlin/Decompose-decompose-js-ir.js:3377:5)\n    at l (webpack-internal:///./kotlin/Decompose-decompose-js-ir.js:3475:18)"[[Prototype]]: RuntimeExceptionconstructor: ƒ IllegalStateException()[[Prototype]]: Exception
protoOf.switch_55pz6e_k$ @ Preconditions.kt:98
protoOf.navigate_tfovv5_k$ @ ChildrenNavigator.kt:108
eval @ ChildrenFactory.kt:130
drainLoop @ Relay.kt:65
protoOf.accept_kmf59w_k$ @ Relay.kt:33
protoOf.navigate_4i2940_k$ @ DefaultStackNavigation.kt:11
navigate @ StackNavigatorExt.kt:7
protoOf.onPopState_a5hf3g_k$ @ DefaultWebHistoryController.kt:127
l @ DefaultWebHistoryController.kt:29
```

</details>

Vivaldi Browser Details:
```
Vivaldi | 6.1.3035.111 (Stable channel) stable (64-bit)
Revision | b2568a9aaecd4a7793d10fa7b389233a3b18bee8
OS | Linux
JavaScript | V8 11.4.183.25
```

Ungoogled Chromium:
```
Version 114.0.5735.106 (Official Build, ungoogled-chromium) (64-bit)
```

If you have Tor Browser, you can surf my app (just a skeleton with navigation setup atm) and see the same issue exists via console logs

```
http://ivf5xtwch5yzptxpbdkjjnxzolln3itw55apkvko3fkdaqdbh6k43jid.onion
```

**Comment by arkivanov** [comment]
Thanks, reproduced. Here are the exact steps:

1. Run the app (the Counters tab is selected)
2. Press tabs in the following order: Cards, Multi-Pane, Dynamic Features (at this point the stack is [Counters, Cards, MultiPane, DynamicFeatures*])
3. Refresh the page (at this point the stack is [DynamicFeatures*])
4. Press Multi-Pane (at this point the stack is [DynamicFeatures, MultiPane*])
5. Press Dynamic Features (at this point the stack is [MultiPane, DynamicFeatures*])
6. Press browser's Back button (at this point the stack is [MultiPane*])
7. Press browser's Forward button (crash)

The issue seems in the experimental `DefaultWebHistoryController`. For some reason at step 7, given the stack is [MultiPane] it tries to push [Cards, MultiPane, MultiPane, DynamicFeatures]. So there are _three_ `MultiPane` configurations in the stack, which is forbidden. This doesn't happen without refreshing the page, so it looks like it messes with the previous session.

Here is the exact line: 

https://github.com/arkivanov/Decompose/blob/0f3b440807a73572f19cd4ef0ba3f3cf512a886f/decompose/src/jsMain/kotlin/com/arkivanov/decompose/router/stack/webhistory/DefaultWebHistoryController.kt#L134

I will try to find the time to check this.

**Comment by arkivanov** [comment]
I think I was able to fix the issue. Will cover with tests and (if no issues found) submit a PR. The issue happens because after page refresh the browser history remains, but the stack is reset. The fix is to restore original stack after page refresh.

**Comment by 05nelsonm** [comment]
Wondering how that might affect my `getInitialStack` strategy when deeplinking

e.g.

```kotlin
fun getInitialStack(deepLink: DeepLink): List<Config> {
    return when (deepLink) {
        is DeepLink.None -> listOf(Config.Home)
        is DeepLink.Web -> {
            when (val config = getConfigForPath(deepLink.path)) {
                // url path: ""
                is Config.Home -> listOf(config)
                // url path: "/next"
                is Config.Next -> listOf(Config.Home, config)
                // url path: "/next/next"
                is Config.NextNext -> listOf(Config.Home, Config.Next, config)
                // ....
            }
        }
    }
}
```

I know the `app-js` sample only calls `getConfigForPath` as the navigation only goes 1 level deep, but do you see the above snippet being an issue with the new functionality of restoring the original stack?

**Comment by arkivanov** [comment]
I think it shouldn't affect much. Most likely, `WebHistoryController` will expose a list of configurations (stack), that could be used as `initialStack` argument. Or your default initial stack if that one is null. But is something I still need to think about.

**Comment by 05nelsonm** [comment]
Maybe an additional parameter to configure `WebHistoryController.attach` to the library consumers' needs?

Something like:

```kotlin
    fun <C : Any> attach(
        purgeHistoryOnPageRefresh: Boolean,
        navigator: StackNavigator<C>,
        stack: Value<ChildStack<C, *>>,
        getPath: (configuration: C) -> String,
        getConfiguration: (path: String) -> C
    )
```

Where if false, the stack is restored from what's in the history and `getConfiguration` is **not** invoked. If true, history is purged and `getConfiguration` **is** invoked.

I know that'd break the api but it's marked as experimental :man_shrugging: ;

**Comment by arkivanov** [comment]
Breaking API is not a problem indeed! An API change would be required any way.

Purging the web history is pretty tricky. E.g. if after page refresh there are two pages in the back history (A and B) and page C is currently shown, and if there is _just one_ child in `Child Stack` (D), then the only way to reset the history is as follows:

1. Assign `window.onPopState` callback 
1. Call `window.history.go(delta = -(historySize - 1))` // Go back to the first page in history (A), all popped pages remain in the forward history (B, C)
2. When `onPopState` callback is called, call `window.history.replaceState` with the only active child (replace A with D)

Unfortunately, calling `replaceState` doesn't erase the forward history. As a result, in this scenario: the currently active page will be D, there will no back history, and the forward history will contain B and C. Which is incorrect.

So It appears that it's better to always restore the stack of `Child Stack` on page refresh, so that the user can consistently go back and forward.

**Comment by arkivanov** [comment]
Another solution could be to implement a Parcelize compiler plugin, similar to Android `kotlin-parcelize` or to `parcelize-darwin` implemented by me for iOS. This would allow saving the entire state (including navigation) to `session storage`, so that it survives page reloads. But this is challenging.

**Comment by arkivanov** [comment]
The fix will be released in the next `2.1.0-alpha-06` version. See the updated [docs](https://arkivanov.github.io/Decompose/navigation/stack/browser-history/).

**Comment by arkivanov** [comment]
Version `2.1.0-alpha-06` is released.
